### ISSUE TITLE
1.0.0-rc.0 pegs CPU and does not finish successfully

### ISSUE THREAD

**Issue by markerikson** [zinserjan/mocha-webpack#163]
I'm in the process of updating a bunch of app build dependencies.  Previously, we were on Webpack 1 and Mocha-Webpack 0.7.  Yesterday, I updated Webpack to 2 and then 3, and saw that Mocha-Webpack 0.7 needed to be bumped.  I updated it to 1.0.0-beta.1, and while I saw some annoying behavior around results display (multiple pages of empty space in the terminal after the results before the tool exited and returned to a prompt), it completed in roughly the same amount of time as 0.7 (on the order of 30-50 seconds).

Today, I saw that 1.0.0-rc.0 was out, so I tried that.  The Mocha-Webpack process successfully runs the Webpack portion, but stays stuck on `[MOCHA} Testing...`.  The Node process keeps a CPU core pegged, and I have yet to see it actually finish and exit - I keep having to kill the process after several minutes.  There's no further output after it shows that it's started testing, either.

I've gone back and forth between 0.7, 1.0.0-beta.1, and 1.0.0-rc.0, and confirmed that beta1 continues to complete okay, while rc0 does not.

Versions:

Node 8.1.2
Webpack 3.3
Mocha-Webpack 1.0.0-rc.0
Windows 7

**Comment by zinserjan** [comment]
> I saw some annoying behavior around results display (multiple pages of empty space in the terminal after the results before the tool exited and returned to a prompt)

Seems that the terminal clearing does not work correctly for you.

> The Node process keeps a CPU core pegged, and I have yet to see it actually finish and exit

Hmm, the only thing that really changed between the rc and the beta was #152. Can you patch mocha-webpack with the previous output of the file and test it again?

**Comment by markerikson** [comment]
Yeah, that's definitely the issue.  I monkey-patched `registerRequireHook.js` to reverse the diff in #152 , and it completes just fine in a few seconds.  If I undo the monkey-patching so that it's back to the code in rc0, it pegs.  Looks like it's an infinite loop in that `for (const resolve of pathResolvers) {` sequence - `resolved` is always `undefined`.

**Comment by zinserjan** [comment]
I'll revert #152 later. The reason why I try to resolve files from memory fs first is that files from memory fs should always take precedence. If a file exists on the real fs and the memory fs, the file from memory is the one we want.

But I think instead of using always the same output path, it's better to use a unique output path.

**Comment by antitoxic** [comment]
Same here. Here's some profiling info after running a single test:
<img width="1679" alt="screen shot 2017-07-18 at 16 40 26" src="https://user-images.githubusercontent.com/261500/28372779-9814a09e-6c6e-11e7-8ec9-88535b80e4e4.png">

**Comment by antitoxic** [comment]
@zinserjan Actually:

1. Why is the resolver used for anything other than the test files? Couldn't it ignore/exclude some dirs by default (node_modules for example).
1. Instead of excluding can `mocha-webpack` only include test files? The files are known after the glob-ing so the list is clear. 

Also I didn't understand this bit:
> But I think instead of using always the same output path, it's better to use a unique output path.

What's going to be the difference in terms of `mocha-webpack`?

**Comment by modosc** [comment]
> 1. Why is the resolver used for anything other than the test files? Couldn't it ignore/exclude some dirs by default (node_modules for example).

sorry to pile on - when upgrading from `0.7` to `1.0` i had to add a `--glob "*.js"` option because otherwise i got loader errors about other files (`.md`, `.json`, even directories). could this be related? happy to open a separate issue if not..

**Comment by zinserjan** [comment]
@antitoxic thanks for the profile screen.

>Why is the resolver used for anything other than the test files? Couldn't it ignore/exclude some dirs by default (node_modules for example).

The idea was to look always on the memory fs first as the file could also exist on the real fs. This is especially useful when you use mocha-webpack with a plugin that writes the files from memory-fs to disc for debugging purposes. 
All output files generated by webpack should be served from memory fs to make sure that we get the latest version of the bundle. This wasn't the case before, but the implementation of memory fs seems to be to inefficient to do this. That's why I reverted #152.

> Instead of excluding can mocha-webpack only include test files? The files are known after the glob-ing so the list is clear

Test files aren't the issue. Output files are the important ones. But I think testing each file path that node tries to resolve against webpack's output path is also a performance killer.  

@markerikson @antitoxic 

mocha-webpack 1.0.0-rc.1 published

@modosc This has nothing to do with this issue. Please open a separate issue with a more detailed description about your problem and also post your webpack config.

**Comment by antitoxic** [comment]
@zinserjan kudos for the quick reaction. 

I'm not sure how `mocha-webpack` patches `mocha` to read files from memory to be honest. I imagined:

1. Globs all test files matching pattern
1. Adds each test file as entry to a single webpack compilation
1. Runs webpack compilation
1. After the first compilation, the list of test files generated in memory is somehow passed to mocha
1. Mocha reads the test files from memory and runs them as if separately.

**Comment by zinserjan** [comment]
@antitoxic almost right, but mocha does not read the files from memory. mocha uses node's module system via require to read and run the tests. That's why need to patch node's require algorithm. Not very pretty but the only solution to get this to work.

**Comment by antitoxic** [comment]
@zinserjan the implementation as it was before - try and catch is very CPU expensive. Every error generates a stack trace. 

If you have (**and `mocha-webpack` definitely has**) a list of files, you can generate a map/hash object with filenames as keys and check against this. This is many times more efficient than try and catch.

I think it's worth a try.
